rasterio/_vsiopener.pyx

```diff
@@ -0,0 +1,154 @@
+# contains path -> PyOpenerBase (or subclass) instances. This is used by
+# the plugin to determine what "files" exist on "disk".
+# Currently the only way to "create" a file in the filesystem is to add
+# an entry to this dictionary. GDAL will then Open the path later.
+cdef _OPENER_REGISTRY = {}
+cdef _OPEN_FILE_OBJS = set()
+
+
+cdef int install_pyopener_plugin(VSIFilesystemPluginCallbacksStruct *callbacks_struct):
+    """Install handlers for python file openers if it isn't already installed."""
+    callbacks_struct = VSIAllocFilesystemPluginCallbacksStruct()
+    callbacks_struct.open = <VSIFilesystemPluginOpenCallback>pyopener_open
+    callbacks_struct.tell = <VSIFilesystemPluginTellCallback>pyopener_tell
+    callbacks_struct.seek = <VSIFilesystemPluginSeekCallback>pyopener_seek
+    callbacks_struct.read = <VSIFilesystemPluginReadCallback>pyopener_read
+    callbacks_struct.close = <VSIFilesystemPluginCloseCallback>pyopener_close
+    callbacks_struct.pUserData = <void*>_OPENER_REGISTRY
+
+    if VSIFileManager.GetHandler("") == VSIFileManager.GetHandler(PREFIX_BYTES):
+        log.debug("Installing Python opener handler plugin...")
+        return VSIInstallPluginHandler(PREFIX_BYTES, callbacks_struct)
+    else:
+        return 0
+
+
+cdef void uninstall_pyopener_plugin(VSIFilesystemPluginCallbacksStruct *callbacks_struct):
+    if callbacks_struct is not NULL:
+        callbacks_struct.pUserData = NULL
+        VSIFreeFilesystemPluginCallbacksStruct(callbacks_struct)
+    callbacks_struct = NULL
+
+
+cdef void* pyopener_open(void *pUserData, const char *pszFilename, const char *pszAccess) with gil:
+    """Access files in the virtual filesystem.
+
+    This function is mandatory in the GDAL Filesystem Plugin API.
+    GDAL may call this function multiple times per filename and each
+    result must be seperately seekable.
+    """
+    if pszAccess != b"r" and pszAccess != b"rb":
+        log.error("Python opener is currently a read-only interface.")
+        return NULL
+
+    if pUserData is NULL:
+        log.error("Python opener registry is not initialized.")
+        return NULL
+
+    cdef dict registry = <object>pUserData
+    filename = pszFilename.decode("utf-8")
+
+    log.debug("Looking up opener: registry=%r, filename=%r", registry, filename)
+    try:
+        file_opener = registry[filename]
+    except KeyError:
+        log.info("Object not found: registry=%r, filename=%r", registry, filename)
+        return NULL
+
+    cdef object file_obj
+
+    try:
+        file_obj = file_opener(filename, "rb")
+    except ValueError:
+        file_obj = file_opener(filename)
+
+    if hasattr(file_obj, "open"):
+        file_obj = file_obj.open()
+
+    log.debug("Opened file object: file_obj=%r", file_obj)
+    _OPEN_FILE_OBJS.add(file_obj)
+    return <void *>file_obj
+
+
+cdef vsi_l_offset pyopener_tell(void *pFile) except -1 with gil:
+    cdef object file_obj = <object>pFile
+    cdef long pos = file_obj.tell()
```

Comment:
I'm not sure if long here really maps to the C type ``long``. But if it does, then you're going to have issues on Windows 64bit where C long is just 32 bit large. Perhaps use ``long long`` or ``uint64_t`` instead. Or just ``vsi_l_offset`` if that's possible

Reply:
Good catch. In cython, `long` is the C `long`. Python's tell() method returns a PyLong object.